File: software/robot/controller/components/utils/PID.ts
//TODO PID class
export class PID {
    kp: number;
    ki: number;
    kd: number;

    first_error: boolean;
    accumulated_error: number;
    previous_error: number;
    previous_error_time: number;


    constructor(kp: number, ki: number, kd: number) {
        this.kp = kp;
        this.ki = ki;
        this.kd = kd;
        this.first_error = true;
        this.previous_error = this.accumulated_error = 0.0;
        this.previous_error_time = Date.now();
    }

    output(error: number): number {
        if (this.first_error) {
            this.first_error = false;
            this.previous_error = error;
            this.previous_error_time = Date.now();
            return this.kp * error;
        } else {
            this.accumulated_error += error * (Date.now() - this.previous_error_time);
            let output: number = this.kp * error +
                this.ki * this.accumulated_error;
                //this.kd * (error - this.previous_error) / (Date.now() - this.previous_error_time);
            this.previous_error = error;
            this.previous_error_time = Date.now();
            return output;
        }
    }
}

File: software/mechanical_computations/PID.py
class PID:
    accumulated_error = 0
    previous_error = 0
    previous_error_time = 0
    first_error = True
    kp = 0
    ki = 0
    kd = 0

    def __init__(self,kp,ki,kd):
        self.kd = kd
        self.ki = ki
        self.kd = kd       
        pass

    def control_variable(self, error, time):
        if self.first_error:
            self.first_error = False
            self.previous_error = error
            self.previous_error_time = time
            return kp * error
        else:
            if time - self.previous_error_time > 0:
                self.accumulated_error = self.accumulated_error + \
                    error * (time - self.previous_error_time)
                output = self.kp * error + self.ki * self.accumulated_error + self.kd * \
                    (error - self.previous_error) / (time - self.previous_error_time)
            else:
                output = self.kp * error + self.ki * self.accumulated_error
            self.previous_error = error
            self.previous_error_time = time
            return output